File: 64. Minimum Path Sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector <vector<int>> T (m+1, vector<int>(n+1));
        for (int i = 1; i <= m; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                if (i == 1) T[1][j] = T[1][j-1] + grid[0][j-1];
                else if (j == 1) T[i][1] = T[i-1][1] + grid[i-1][0];
                else
                    T[i][j] = T[i-1][j] + grid[i-1][j-1] < T[i][j-1] + grid[i-1][j-1] ? T[i-1][j] + grid[i-1][j-1] : T[i][j-1] + grid[i-1][j-1];
            }
        }
        return T[m][n];
    }
};

File: 64. Minimum Path Sum.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m, n = len(grid), len(grid[0])
        T = [[0 for j in xrange(n + 1)] for i in xrange(m + 1)]
        for i in xrange(1, m + 1):
            for j in xrange(1, n + 1):
                if i == 1:
                    T[1][j] = T[1][j-1] + grid[0][j-1]
                elif j == 1:
                    T[i][1] = T[i-1][1] + grid[i-1][0]
                else:
                    T[i][j] = min(T[i-1][j], T[i][j-1]) + grid[i-1][j-1]
        return T[m][n]
